docs/timeouts.rst

```diff
@@ -0,0 +1,180 @@
+Timeouts
+========
+
+.. currentmodule:: trio_websocket
+
+Networking code is inherently complex due to the unpredictable nature of network
+failures and the possibility of a remote peer that is coded incorrectly—or even
+maliciously! Therefore, your code needs to deal with unexpected circumstances.
+One common failure mode that you should guard against is a slow or unresponsive
+peer.
+
+This page describes the timeout behavior in ``trio-websocket`` and shows various
+examples for implementing timeouts in your own code. Before reading this, you
+might find it helpful to read `"Timeouts and cancellation for humans"
+<https://vorpus.org/blog/timeouts-and-cancellation-for-humans/>`__, an article
+written by Trio's author that describes an overall philosophy regarding
+timeouts. The short version is that Trio discourages libraries from using
+internal timeouts. Instead, it encourages the caller to enforce timeouts, which
+makes timeout code easier to compose and reason about.
+
+On the other hand, this library is intended to be safe to use, and omitting
+timeouts could be a dangerous flaw. Therefore, this library contains takes a
+balanced approach to timeouts, where high-level APIs have internal timeouts, but
+you may disable them or use lower-level APIs if you want more control over the
+behavior.
+
+Built-in Client Timeouts
+------------------------
+
+The high-level client APIs :func:`open_websocket` and :func:`open_websocket_url`
+contain built-in timeouts for connecting to a WebSocket and disconnecting from a
+WebSocket. These timeouts are built-in for two reasons:
+
+1. Omitting timeouts may be dangerous, and this library strives to make safe
+   code easy to write.
+2. These high-level APIs are context managers, and composing timeouts with
+   context managers is tricky.
+
+These built-in timeouts make it easy to write a WebSocket client that won't hang
+indefinitely if the remote endpoint or network are misbehaving. The following
+example shows a connect timeout of 10 seconds. This guarantees that the block
+will start executing (reaching the line that prints "Connected") within 10
+seconds. When the context manager exits after the ``print(Received response:
+…)``, the disconnect timeout guarantees that it will take no more than 5 seconds
+to reach the line that prints "Disconnected". If either timeout is exceeded,
+then the entire block raises ``trio.TooSlowError``.
+
+.. code-block:: python
+
+    async with open_websocket_url('ws://my.example/', connect_timeout=10,
+            disconnect_timeout=5) as ws:
+        print("Connected")
+        await ws.send_message('hello from client!')
+        response = await ws.get_message()
+        print('Received response: {}'.format(response))
+    print("Disconnected")
+
+.. note::
+
+    The built-in timeouts do not affect the contents of the block! In this
```

Comment:
Are you objecting to the phrase "affect the contents" or the term "block"? 

I think the former is accurate, in the sense that nothing you do inside that block will be affected by these timeout arguments, i.e. even if you `trio.sleep(math.inf)` that wouldn't time out. It would be overly narrow to say that the timeouts don't affect messaging.

As for "block", I'm uncertain. I originally wrote something more verbose like "context manager block" but it felt unwieldy to keep using that phrase. Maybe spell it out once and then abbreviate it as CM block?